File: implementations/ts/src/alias.ts
// Relation signatures (SPEC-9 §5): the deterministic answer to "what relation shape does this
// delta instantiate?" — the librarian's input, never part of evaluation semantics.

import { array, encode, tstr } from "./cbor.js";
import { bytesToHex } from "./hash.js";
import type { Delta } from "./types.js";

function byteCompare(a: Uint8Array, b: Uint8Array): number {
  const n = Math.min(a.length, b.length);
  for (let i = 0; i < n; i++) {
    const d = a[i]! - b[i]!;
    if (d !== 0) return d;
  }
  return a.length - b.length;
}

// The [role, context] pairs ([role] when the pointer has no context) of the delta's EntityRef
// pointers, sorted bytewise by their canonical CBOR encoding. Primitive and DeltaRef pointers
// contribute nothing: primitives are not vertices (SPEC-1 §2.3); delta references are plumbing.
export function relationSignature(delta: Delta): readonly (readonly string[])[] {
  const pairs: { pair: string[]; bytes: Uint8Array }[] = [];
  for (const ptr of delta.claims.pointers) {
    if (ptr.target.kind !== "entity") continue;
    const context = ptr.target.entity.context;
    const pair = context === undefined ? [ptr.role] : [ptr.role, context];
    pairs.push({ pair, bytes: encode(array(pair.map(tstr))) });
  }
  pairs.sort((a, b) => byteCompare(a.bytes, b.bytes));
  return pairs.map((p) => p.pair);
}

// The signature's canonical form: the canonical CBOR of the sorted array of pairs.
export function relationSignatureCanonicalHex(delta: Delta): string {
  const pairs = relationSignature(delta).map((pair) => array(pair.map(tstr)));
  return bytesToHex(encode(array(pairs)));
}

File: implementations/rust/src/alias.rs
//! Relation signatures (SPEC-9 §5). Mirrors ../ts/src/alias.ts — the deterministic answer to
//! "what relation shape does this delta instantiate?": the librarian's input, never part of
//! evaluation semantics.

use crate::cbor::{encode, CborValue};
use crate::types::{Delta, Target};

/// The [role, context] pairs ([role] when the pointer has no context) of the delta's EntityRef
/// pointers, sorted bytewise by their canonical CBOR encoding. Primitive and DeltaRef pointers
/// contribute nothing: primitives are not vertices (SPEC-1 §2.3); delta references are plumbing.
pub fn relation_signature(delta: &Delta) -> Vec<Vec<String>> {
    let mut pairs: Vec<(Vec<u8>, Vec<String>)> = delta
        .claims
        .pointers
        .iter()
        .filter_map(|ptr| {
            let Target::Entity(er) = &ptr.target else {
                return None;
            };
            let mut pair = vec![ptr.role.clone()];
            if let Some(c) = &er.context {
                pair.push(c.clone());
            }
            let bytes = encode(&CborValue::Array(
                pair.iter().map(|s| CborValue::Tstr(s.clone())).collect(),
            ));
            Some((bytes, pair))
        })
        .collect();
    pairs.sort_by(|a, b| a.0.cmp(&b.0));
    pairs.into_iter().map(|(_, pair)| pair).collect()
}

/// The signature's canonical form: the canonical CBOR of the sorted array of pairs.
pub fn relation_signature_canonical_hex(delta: &Delta) -> String {
    let arrays: Vec<CborValue> = relation_signature(delta)
        .into_iter()
        .map(|pair| CborValue::Array(pair.into_iter().map(CborValue::Tstr).collect()))
        .collect();
    hex::encode(encode(&CborValue::Array(arrays)))
}
